C# 7.x: out Variables

RexJaeschke opened the pull request:
The handling of discards in tuple deconstruction and pattern matching is deferred to the PRs that add support for tuples and pattern matching, respectively. This PR only handles the out parameter and assignment scenarios.

A couple of `> TODO` placeholders will need attention once those other PRs are created.

RexJaeschke committed:
Update lexical-structure.md

RexJaeschke committed:
Update variables.md

RexJaeschke committed:
Update expressions.md

RexJaeschke committed:
Update classes.md

RexJaeschke committed:
Fixup link for local variable scope to new patterns subclause

Part of adding V7.3 support for out variables in extra places.

BillWagner committed:
Fix markdown lint errors, which included a missed conflict

gafter commented on standard/lexical-structure.md:
```diff
@@ -64,18 +64,17 @@ The productions for *simple_name* ([§11.7.4](expressions.md#1174-simple-names))
 > *end example*
 
-If a sequence of tokens can be parsed (in context) as a *simple_name* ([§11.7.4](expressions.md#1174-simple-names)), *member_access* ([§11.7.6](expressions.md#1176-member-access)), or *pointer_member_access* ([§22.6.3](unsafe-code.md#2263-pointer-member-access)) ending with a *type_argument_list* ([§8.4.2](types.md#842-type-arguments)), the token immediately following the closing `>` token is exa ...
+If a sequence of tokens can be parsed (in context) as a *simple_name* ([§11.7.4](expressions.md#1174-simple-names)), *member_access* ([§11.7.6](expressions.md#1176-member-access)), or *pointer_member_access* ([§22.6.3](unsafe-code.md#2263-pointer-member-access)) ending with a *type_argument_list* ([§8.4.2](types.md#842-type-arguments)), the token immediately following the closing `>` token is exa ...
```
This disambiguation is not correct.  The correct disambiguation rule appears in #61.  This rule improperly changes the meaning of `M(a < b > c)` from its current meaning (a syntactically valid chain of comparison operators) into a syntax error.  See [HERE](https://sharplab.io/#v2:CYLg1APgAgTABAQQLACgDeq5blAzDgFjgFkBDASwDsAKASk2wxWxbgA05S4BeOSgUwDu7OgG4GrdnABGPTuOaSpAYzmkFSktS4AeGXAB8cZbQ3YAvhMJaOAD1pw0cSyhepY7VExZ4cARgA2KQB7AAd+ACdSABdgiLgdag4AG34AM2iAGikI8gBzAAtoh24jVIyzLF8oQJDwqNj4gyS4cqyc/KKSsvTohRcgA) for a test that demonstrates that this is not correct.

BillWagner replied:
See https://github.com/dotnet/csharpstandard/pull/61/files#r869861306

BillWagner replied:
Brought in the changes from #61 (which I believe can be closed without merging).

RexJaeschke committed:
undo earlier edits

This text has been moved to PR 596.

RexJaeschke committed:
undo previous edits

Some of this text has been moved to PR 596

RexJaeschke committed:
add new conversion

RexJaeschke committed:
undo earlier edits

- Overhauled argument list text
- Added better conversion text
- Removed discards-in-assignment text

RexJaeschke committed:
undo earlier edits

RexJaeschke committed:
var-ref becomes identifier

RexJaeschke committed:
fix spacing

RexJaeschke committed:
fix spacing

jcouv commented on standard/expressions.md:
```diff
@@ -541,7 +541,7 @@ argument_name
     : expression
     | 'ref' variable_reference
-    | 'out' variable_reference
+    | 'out' local_variable_type? identifier
```
The speclets for out vars and deconstructions relied on `declaration_expression`.
But in this doc, I see `'out' local_variable_type? identifier` and in the deconstruction [draft](https://github.com/dotnet/csharpstandard/pull/63/files#diff-d9d678d0b42d3c668ebd64159dadfd6188361121feca8fb94fddf8816433cd14R6357) I see `tuple_deconstruction_expression`.
I'm not sure what is the benefit of this structure.

I think having a `declaration_expression` that appears in multiple contexts is a good factorization. Then we don't need new syntax for deconstructions (it's `declaration_expression` and `tuple_expression`).

Tagging @cston since I'll be away.

jcouv replied:
Some more context on syntax shape (was discussed with LDM): https://github.com/dotnet/csharpstandard/issues/285

BillWagner replied:
Updated to use `declaration_expression`.

Tagging @Nigel-Ecma for grammar checks.

Nigel-Ecma replied:
*declaration_expression* is different and cannot simply be substituted in this location. Here *local_variable_type* is optional, it is not in *declaration_expression*. I'll find the change and add a suggestion there…

BillWagner commented on standard/expressions.md:
```diff
@@ -553,7 +553,55 @@ The *argument_value* can take one of the following forms:
 - An *expression*, indicating that the argument is passed as a value parameter ([§14.6.2.2](classes.md#14622-value-parameters)).
 - The keyword `ref` followed by a *variable_reference* ([§9.5](variables.md#95-variable-references)), indicating that the argument is passed as a reference parameter ([§14.6.2.3](classes.md#14623-reference-parameters)). A variable shall be definitely assigned ([§9.4](variables.md#94-definite-assignment)) before it can be passed as a reference parameter.
-- The keyword `out` followed by a *variable_reference* ([§9.5](variables.md#95-variable-references)), indicating that the argument is passed as an output parameter ([§14.6.2.4](classes.md#14624-output-parameters)). A variable is considered definitely assigned ([§9.4](variables.md#94-definite-assignment)) following a function member invocation in which the variable is passed as an output parameter ...
+- The keyword `out`, optionally followed by *local_variable_type* ([§12.6.2](statements.md#1262-local-variable-declarations)), followed by *identifier*, indicating that the argument is passed as an output parameter ([§14.6.2.4](classes.md#14624-output-parameters)). If *identifier* is not a discard (§discards-new-clause), the variable it designates is considered definitely assigned ([§9.4](variabl ...
```
This is where the build warnings are coming from. The discards clause is not part of this PR.

BillWagner committed:
fix build warning

BillWagner committed:
fix build warnings

BillWagner committed:
Final edits.

BillWagner commented:
@jskeet This one is now ready for the upcoming meeting.

Latest edits:

- rebase
- update all links
- remove any discard related text (that has been done in a separate PR.)

Nigel-Ecma commented on standard/expressions.md:
```diff
@@ -549,7 +549,7 @@ argument_name
     : expression
     | 'ref' variable_reference
-    | 'out' variable_reference
+    | 'out' declaration_expression? identifier
```
While the idea of factoring out to a common shared production is not wrong *per se* in this case doing so has changed the grammar ☹️. Revert to the intermediate version:

```suggestion
    | 'out' local_variable_type? identifier
```

jskeet replied:
Other than generally trusting Nigel on matters of grammar, I don't have an opinion on this one - but it needs to be resolved one way or another.

Nigel-Ecma reviewed (changes requested):
I haven't done a full review just looked at the grammar change as requested – unfortunately the change was wrong so needs reverting, suggestion in comment.

jskeet commented on standard/expressions.md:
```diff
@@ -559,7 +559,55 @@ The *argument_value* can take one of the following forms:
+
+- If *local_variable_type* is present
+  - If *identifier* is `_`, it is interpreted as a typed discard
+  - Otherwise, *identifier* denotes a new local variable. It is a compile-time error if a variable by that name is currently in scope or if there is not an implicit conversion between the corresponding parameter’s type and *local_variable_type*.
```
Two problems with this:

- It doesn't specify the direction of the required conversion (from the parameter type to the local variable type, or vice versa?)
- I suspect it actually needs to be an identity conversion.

Example:

```csharp
class Program
{
    static void Main()
    {
        // Error: Argument 1: cannot convert from 'out object' to 'out string'
        M(out object x);
    }

    static void M(out string y) => y = "";
}
```

BillWagner replied:
Correct. Verified that an identity conversion is required.

jskeet commented on standard/expressions.md:
````diff
@@ -559,7 +559,55 @@ The *argument_value* can take one of the following forms:
+>
+> ```csharp
+> static void M(out int x) { … }
+> static void CallM1()
````
Add a blank line between the methods?

jskeet commented on standard/expressions.md:
```diff
@@ -559,7 +559,55 @@ The *argument_value* can take one of the following forms:
+> static void M(out int x) { … }
+> static void CallM1()
+> {
+>     int not_;
```
`not_` feels like an odd variable name to use for this example. Can we just call it `v` or something similarly neutral?

BillWagner replied:
updated

jskeet commented on standard/expressions.md:
```diff
@@ -559,7 +559,55 @@ The *argument_value* can take one of the following forms:
+>     M(out uint pu);   // error: new variable but has incompatible type
+>     M(out int _);     // OK: typed discard
+>     M(out var not_3); // OK: typed new variable
+>     M(out var _);     // OK: typed discard
```
"implicitly typed discard" to differentiate between this and line 582? Not sure - let's discuss.

BillWagner replied:
Updated. Leaving unresolved, but I believe you are correct.

jskeet commented on standard/expressions.md:
```diff
@@ -559,7 +559,55 @@ The *argument_value* can take one of the following forms:
+> *end example*
+
+- If *local_variable_type* is absent
+  - If *identifier* is `_`
```
I think we need an "otherwise" for this "if".

jskeet commented on standard/expressions.md:
```diff
@@ -559,7 +559,55 @@ The *argument_value* can take one of the following forms:
+>
+> *end example*
+
+It is also an error to reference an implicitly-typed out variable in the same argument list that immediately contains its declaration. The type of an implicitly-typed out variable is the type of the corresponding parameter in the signature of the method selected by overload resolution.
```
"It is also an error" seems a little odd after two examples. Just "It is an error"? I think an example of this would be useful, too.

Also the second sentence of this paragraph seems unrelated to the first. Separate paragraph?

jskeet commented on standard/expressions.md:
```diff
@@ -1033,6 +1081,8 @@ Given an implicit conversion `C₁` that converts from an expression `E` to a ty
 - `E` exactly matches both or neither of `T₁` and `T₂`, and `T₁` is a better conversion target than `T₂` ([§12.6.4.6](expressions.md#12646-better-conversion-target))
 - `E` is a method group ([§12.2](expressions.md#122-expression-classifications)), `T₁` is compatible ([§20.4](delegates.md#204-delegate-compatibility)) with the single best method from the method group for conversion `C₁`, and `T₂` is not compatible with the single best method from the method group for conversion `C₂`
 
+The  conversion from an implicitly-typed out variable declaration is not considered better than any other conversion from expression.
```
While I know what this means, it doesn't feel terribly clear - maybe we just need a reference to the "conversion from expression" section?

BillWagner replied:
I'm also not sure it adds anything. Should this paragraph be removed? For discussion.

jskeet commented on standard/lexical-structure.md:
```diff
@@ -65,7 +65,7 @@ The productions for *simple_name* ([§12.8.4](expressions.md#1284-simple-names))
 > *end example*
 
-If a sequence of tokens can be parsed (in context) as a *simple_name* ([§12.8.4](expressions.md#1284-simple-names)), *member_access* ([§12.8.7](expressions.md#1287-member-access)), or *pointer_member_access* ([§23.6.3](unsafe-code.md#2363-pointer-member-access)) ending with a *type_argument_list* ([§8.4.2](types.md#842-type-arguments)), the token immediately following the closing `>` token is exa ...
+If a sequence of tokens can be parsed (in context) as a *simple_name* ([§12.8.4](expressions.md#1284-simple-names)), *member_access* ([§12.8.7](expressions.md#1287-member-access)), or *pointer_member_access* ([§23.6.3](unsafe-code.md#2363-pointer-member-access)) ending with a *type_argument_list* ([§8.4.2](types.md#842-type-arguments)), the token immediately following the closing `>` token is exa ...
```
I don't think this really works with how the continuation is described. I'd revert this, *or* change the text at line 75.

(It's not immediately clear why this is in an `out` variable PR anyway, mind you.)

BillWagner replied:
That was trying to fix a merge conflict. Reverted.

jskeet commented on standard/expressions.md:
```diff
@@ -6051,7 +6101,7 @@ assignment_operator
 The left operand of an assignment shall be an expression classified as a variable, a property access, an indexer access, an event access or a tuple. A declaration expression is not directly permitted as a left operand, but may occur as a step in the evaluation of a deconstructing assignment.
 
-The `=` operator is called the ***simple assignment operator***. It assigns the value or values of the right operand to the variable, property, indexer element or tuple elements given by the left operand. The left operand of the simple assignment operator shall not be an event access (except as described in [§15.8.2](classes.md#1582-field-like-events)). The simple assignment operator is described ...
+The `=` operator is called the ***simple assignment operator***. When the left operand is not a discard (§9.2.8.1), this operator assigns the value of the right operand to the variable, property, or indexer element given by the left operand. Otherwise, the right operand is evaluated with its value going unused. The left operand of the simple assignment operator shall not be an event access (excep ...
```
Is this change and the next one intended to be in this PR? It's not really about `out` parameters... was it just something else that should have been in a discards PR?

BillWagner replied:
That's correct. This PR in its original form included both discards and out parameters. I left this change in place because it wasn't in the updated discards and tuples PR.

Let's discuss at the meeting if it should be removed.

jskeet replied:
Action from 2023-04-19: @BillWagner to remove from this PR, but add it to the issue that @MadsTorgersen has already created for polishing on discards.

jskeet commented on standard/lexical-structure.md:
```diff
@@ -77,7 +77,8 @@ If the following token is among this list, or an identifier in such a context, t
 <!-- markdownlint-disable MD028 -->
 
-<!-- markdownlint-enable MD028 -->
+> *Note*: In an *argument_value* of the form `out local_variable_type? variable_reference`, *local_variable_type* may be a generic type; for example: `F(out A<B> name)`. As the language grammar for the argument uses *variable_reference* (which is really *expression*), this context is subject to the disambiguation rule. In this case the closing `>` is followed by an identifier. *end note*
```
This is inaccurate regardless of whether we use Nigel's suggestion - the grammar *used* to use *variable_reference* but it doesn't any more.

BillWagner replied:
This was also added when I was resolving merge conflicts. It could be reverted back. Let's discuss at the meeting.

Nigel-Ecma replied:
I don't recall being responsible for the change from *variable_reference* to *identifier* but whoever it was I believe they are correct. The change does seem to make the note a bit pointless as it is explaining that the *variable_reference* can only be an *identifier* and so the disambiguation rules will apply and this will parse correctly…

However, a revised note might be:

```suggestion
> *Note*: In an *argument_value* of the form `out local_variable_type? identifier`, *local_variable_type* may be a generic type; for example: `F(out A<B> name)`. This example would be subject to the disambiguation rule ([§6.2.5](lexical-structure.md#625-grammar-ambiguities)) and is resolved correct;y as the closing `>` is followed by an identifier. *end note*
```

(Check my x-ref syntax before accepting!)

BillWagner replied:
This xref looks good.

jskeet replied:
Action from 2023-04-19: @BillWagner  to remove the note.

BillWagner committed:
Respond to pre-meeting feedback.

jskeet commented on standard/expressions.md:
```diff
@@ -549,7 +549,7 @@ argument_name
     : expression
     | 'ref' variable_reference
-    | 'out' variable_reference
+    | 'out' declaration_expression? identifier
```
Action from 2023-04-19:

```suggestion
    | 'out' declaration_expression
    | 'out' variable_reference
```

Reasoning: declaration_expression always has a local variable type, whereas it's fine to use an argument_value of `out x` if `x` is already declared, for example.

jskeet reviewed (approved):
Some simple final changes, then let's merge.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Jon Skeet <jonskeet@google.com>
Co-authored-by: Nigel-Ecma <perryresearch@zoot.net.nz>

BillWagner committed:
feedback from April meeting

MadsTorgersen commented:
To the best of my analysis, this PR is fully replaced by #664.

jskeet commented:
Closing as obsolete.
